### DISCUSSION TITLE
Schema alias issue

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#133] [Design & Development]
I was revising the Metadata tutorial and ran across something that is either a gap in the implementation of Schema or something you (Ian) may have decided to omit.  

In MetadataDefinitions we had this method:
```
  /*! \brief Apply a set of aliases to data.

  This method should be called in processing workflows to apply a series
  of defined aliases to data.  The method uses the is_alias method to
  verify if an alias name is valid.  It will not change the key if the name is
  not defined as a valid alias.   The keys that fail that test will be
  posted to the std::list that is returned.   Callers should test the
  size of the return and if it is not empty take appropriate action.

    \param d is the data to alter (usually actually a TimeSeries of Seismogram)
    \param aliaslist is a list of aliases names to apply.

    \return std::list of srings of failed changes.  Callers should
     test the size of this return and take action if needed.
    */
std::list<std::string> apply_aliases(mspass::utility::Metadata& d,
          const std::list<std::string> aliaslist);
```
The idea of this method was to automate the process of applying a standard set of aliases like obspy or SAC.  That method is not found in schema.py.  In fact, I just realized it is noted in the file as "TODO".   

What is the plan for that?  We might want to consider a different design than the C++ code.   Seems to me a useful signature for the implementation in python would be this:
```
def apply_aliases(self,md,alias_name='SAC',alias_to_apply=None):
```
md is the Metadata container to apply the aliases to, *alias_name* would refer to a defined name set like SAC and *alias_to_apply* would be a dict (might prefer a list of 2 element tuples) of a custom set of aliases to apply.  The later would provide a mechanism to define aliases on the fly without having to define the aliases in the schema.  That functionality might be a problem thought.  Not sure enough about what is cached in the current schema.py implementation.

**Comment by wangyinz** [comment]
I just had that implemented in https://github.com/wangyinz/mspass/pull/131/commits/89adaf397d4018525ec7a275f3b4861a4e21e0f8. It uses dict as the input and will automatically reference of all the aliases in its internal container, so users can call the clear_alias method from the same schema object to reverse the change. I think that should work.

**Reply by pavlis** [reply]
Cool, useful addition.
